### PR TITLE
add predictoor template

### PR DESCRIPTION
**Pull request by alexcos20**
For [this epic](https://github.com/oceanprotocol/predictoor-pm/issues/4) "[Predictoor BE] Ship MVP" -- Add DT3 to contracts repo

Discussion on whether to merge to main: [predictoor-pm#18](https://github.com/oceanprotocol/predictoor-pm/issues/18)

### CODE REVIEW

File: contracts/templates/ERC20TemplatePredictoor.sol [hunk 1]
```diff
@@ -0,0 +1,1155 @@
+pragma solidity 0.8.12;
```
Comment (trentmc) [inline contracts/templates/ERC20TemplatePredictoor.sol:1]: I'd like a better name than `ERC20TemplatePredictoor.sol`. (As we discussed).

Datapoint: We will likely have _more_ predictoor templates over time.
- Built-in oracle
- Different formula to compensate predictoors
- Bug fixes

Datapoint: `address.json` refers to templates by their index. 1 is ERC20Template, 2 is ERC20TemplateEnterprise, etc. As this list grows then it'll be harder to cross-reference. Ref:
```text
    "ERC20Template": {
      "1": "0x9b87E416a76d8bCAf59A5c746D5ceB57119F08Cb",
      "2": "0x12bB8D85a091A69A07E22E52d4567dBB91568f52",
      "3": "0x9497d1d64F2aFeBcd4f9916Eef3d9094E5Df962f"
    },
```

Datapoint: `ERC20TemplateEnterprise` is not a great name. Perhaps the original driver was an enterprise customer, but there are other potential users. The name has led to great confusion about where `ERC20TemplateEnterprise` is applicable.

[This comment](https://github.com/oceanprotocol/ocean.py/issues/1315#issuecomment-1432848930) gives further discussion and analysis of names. To go forward, it proposes a solution too: (a) have a _table_ that identifies the distinguishing characteristics of each template; and (b) from now on, label the templates ERC20Template3.sol, ERC20Template4.sol, etc.

Given the above - and as discussed - could we name _this_ new template "ERC20Template3.sol" please?

Comment (trentmc) [reply contracts/templates/ERC20TemplatePredictoor.sol:1]: Update: @alexcos20 created an issue to create this table

https://github.com/oceanprotocol/contracts/issues/754

File: contracts/templates/ERC20Template3.sol [hunk 2]
```diff
@@ -0,0 +1,1184 @@
...
+    }
+
+    function get_subscription_revenue_at_block(
+        uint256 blocknum
+    ) public view returns (uint256) {
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        return (subscription_revenue_at_block[slot]);
+    }
+
+    function get_prediction(
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:978]: Rename to `get_predval()`

Why: to disambiguate from each predictoor's `predval`, versus the top-level aggregated predval `agg_predval`.

File: contracts/templates/ERC20Template3.sol [hunk 3]
```diff
@@ -0,0 +1,1184 @@
...
+    uint256 private rate = 0;
+    uint256 public constant BASE = 1e18;
+
+    // -------------------------- PREDICTOOR --------------------------
+    enum Status {
+        Pending,
+        Paying,
+        Canceled
+    }
+    event PredictionSubmitted(
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:56]: Rename to `PredvalSubmitted`

Why: disambiguate from each predictoor's `predval`, versus the top-level aggregated predval `agg_predval`.

File: contracts/templates/ERC20Template3.sol [hunk 4]
```diff
@@ -0,0 +1,1184 @@
...
+        address indexed predictoor,
+        uint256 indexed slot,
+        uint256 stake
+    );
+    event PredictionPayout(
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:61]: Rename to `PredictoorPayout`

Why: more specific / clearer

File: contracts/templates/ERC20Template3.sol [hunk 5]
```diff
@@ -0,0 +1,1184 @@
...
+        address indexed predictoor,
+        uint256 indexed slot,
+        uint256 stake,
+        uint256 payout,
+        bool prediction,
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:66]: Rename to `predval`

Why: disambiguate from each predictoor's `predval`, versus the top-level aggregated predval `agg_predval`.

File: contracts/templates/ERC20Template3.sol [hunk 6]
```diff
@@ -0,0 +1,1184 @@
...
+        bool truval,
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:67]: Rename to `trueval`

Why: fix spelling mistake

File: contracts/templates/ERC20Template3.sol [hunk 7]
```diff
@@ -0,0 +1,1184 @@
...
+        // if OPF hasn't submitted trueval in truval_submit_timeout blocks then cancel round
+        if (block.number > slot + truval_submit_timeout_block && truval_submitted[slot]==Status.Pending){
+            truval_submitted[slot]=Status.Canceled;
+        }
+
+        if(truval_submitted[slot]==Status.Pending){
+            // if Status is Pending, do nothing, just return
+            return; 
+        }
+        if(truval_submitted[slot]==Status.Canceled){
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:1047]: There are three separate constructions of the event `PredictionPayout`. It makes this routine hard to follow.

Could we make this cleaner please?
- Idea 1: just create one event object at the top, with initial params. Then change its params a bit depending on each of the three scenarios. Then emit.
- Idea 2: other?

File: contracts/templates/ERC20Template3.sol (hunk 6 above)
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:67]: Maybe: also emit `agg_predval`, if convenient.

Why: will be helpful in debugging, to have side-by-side with this agent's `predval` and the `trueval`.

Only do if convenient.

File: contracts/templates/ERC20Template3.sol [hunk 8]
```diff
@@ -0,0 +1,1184 @@
...
+        address indexed user,
+        uint256 expires,
+        uint256 blocknum
+    );
+    event TruevalSubmitted(
+        uint256 indexed slot,
+        bool trueval,
+        Status status
+    );
+    struct Prediction {
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:80]: I wonder if we should rename this to `PredictionStruct`? Will help to disambiguate from `predval` which it holds.

File: contracts/templates/ERC20Template3.sol [hunk 9]
```diff
@@ -0,0 +1,1184 @@
...
+
+    struct fixedRate {
+        address contractAddress;
+        bytes32 id;
+    }
+    fixedRate[] fixedRateExchanges;
+    address[] dispensers;
+
+    struct providerFee {
+        address providerFeeAddress;
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:131]: Give a comment. E.g. `// address where to send collected fees`

File: contracts/templates/ERC20Template3.sol [hunk 10]
```diff
@@ -0,0 +1,1184 @@
...
+        address providerFeeToken; // address of the token
```
Comment (trentmc) [inline contracts/templates/ERC20Template3.sol:132]: Change comment to: `// token which provider fees are denominated in`

Why: clarity

**Review by trentmc — CHANGES_REQUESTED** [review]
More comments. More to come (on Mon:)

File: contracts/templates/ERC20Template3.sol (hunk 2 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:978]: This returns entire prediction (predval,stake,predictoor,paid) and not just the predval

File: contracts/templates/ERC20Template3.sol (hunk 3 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:56]: we don't have the predval in the event (would be easy for somebody else to see it), we only have the information that a prediction for slot X has been submited by predictoor Y and he/she staked Z

File: contracts/templates/ERC20Template3.sol (hunk 4 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:61]: A Predictoor can have multiple payouts.   It's more clear "Prediction" because then we know that it refers to a certain prediction

File: contracts/templates/ERC20Template3.sol (hunk 6 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:67]: done

File: contracts/templates/ERC20Template3.sol (hunk 9 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:131]: added in code:
```
// this structure is here only for compatibility reasons with other datatoken templates
// it's not validated or used anywhere in the code, except as unused argument to startOrder function
    ```

File: contracts/templates/ERC20Template3.sol (hunk 10 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:132]: see above

File: contracts/templates/ERC20Template3.sol (hunk 7 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:1047]: unfortunately, contract size is 23.971 Kb.
Adding a couple of variables to store event params will make it > 24K, making it undeployable :(

File: contracts/templates/ERC20Template3.sol (hunk 8 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:81]: This is a prediction.
A prediction is formed of :
 - predictoor address
 - predval
 - stake
 - paid   (if it was paid or not)

File: contracts/templates/ERC20Template3.sol (hunk 7 above)
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:1027]: solved

**Comment by alexcos20** [comment]
TODO:  use camelCase every where

**Review by trentmc — CHANGES_REQUESTED** [review]
